cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -79,16 +88,42 @@ class ParquetScanTask : public ScanTask {
       std::unique_ptr<RecordBatchReader> record_batch_reader;
     } NextBatch;
 
+    RETURN_NOT_OK(EnsurePreBuffered());
     NextBatch.file_reader = reader_;
     RETURN_NOT_OK(reader_->GetRecordBatchReader({row_group_}, column_projection_,
                                                 &NextBatch.record_batch_reader));
     return MakeFunctionIterator(std::move(NextBatch));
   }
 
+  // Ensure that pre-buffering has been applied to the underlying Parquet reader
+  // exactly once (if needed). If we instead set pre_buffer on in the Arrow
+  // reader properties, each scan task will try to separately pre-buffer, which
+  // will lead to crashes as they trample the Parquet file reader's internal
+  // state. Instead, pre-buffer once at the file level. This also has the
+  // advantage that we can coalesce reads across row groups.
+  Status EnsurePreBuffered() {
+    if (pre_buffer_once_) {
+      BEGIN_PARQUET_CATCH_EXCEPTIONS
+      std::call_once(*pre_buffer_once_, [this]() {
+        reader_->parquet_reader()->PreBuffer(pre_buffer_row_groups_, column_projection_,
+                                             async_context_, cache_options_);
+      });
+      END_PARQUET_CATCH_EXCEPTIONS
+    }
+    return Status::OK();
+  }
+
  private:
   int row_group_;
   std::vector<int> column_projection_;
   std::shared_ptr<parquet::arrow::FileReader> reader_;
+  // Pre-buffering state. pre_buffer_once will be nullptr if no
+  // pre-buffering is to be done, and row_groups will be empty in that
+  // case. We assume all scan tasks have the same column projection.
```

Comment:
Not really a problem, butyou still set row_groups even if when not pre-buffering